lib/App/Config/Chronicle.pm

```diff
@@ -347,6 +350,316 @@ sub current_revision {
+
+    return 1;
+}
+
+sub _has_refresh_interval_passed {
+    my $self                   = shift;
+    my $now                    = time;
+    my $prev_update            = $self->_updated_at;
+    my $time_since_prev_update = $now - $prev_update;
+    return ($time_since_prev_update > $self->refresh_interval);
+}
+
+sub _is_cache_stale {
+    my $self      = shift;
+    my @rev_cache = $self->_retrieve_objects_from_cache(['_global_rev']);
+    my @rev_chron = $self->_retrieve_objects_from_chron(['_global_rev']);
+    return !($rev_cache[0] && $rev_chron[0] && $rev_cache[0]->{data} eq $rev_chron[0]->{data});
+}
+
+=head2 global_revision
+
+Returns the global revision version of the config chronicle.
+This will correspond to the last time any of values were changed.
+
+=cut
+
+sub global_revision {
+    my $self = shift;
+    return ($self->get('_global_rev'));
+}
+
+=head2 set
+
+Takes a hashref of key->value pairs and atomically sets them in config chronicle
+
+Example:
+    set({key1 => 'value1', key2 => 'value2', key3 => 'value3',...});
+
+=cut
+
+sub set {
+    my ($self, $pairs) = @_;
+    my $rev_obj   = Date::Utility->new;
+    my $rev_epoch = $rev_obj->{epoch};
+
+    grep { die "Cannot set with key: $_ | Key must be defined with 'global: 1'" unless $self->_key_is_dynamic($_) } keys %$pairs;
+
+    $pairs->{_global_rev} = $rev_epoch;
+    my %key_objs_hash = pairmap { $a => {data => $b, _local_rev => $rev_epoch} } %$pairs;
+    $self->_store_objects(\%key_objs_hash, $rev_obj);
+
+    return 1;
+}
+
+sub _store_objects {
+    my ($self, $key_objs_hash, $date_obj, $optional_chron_args) = @_;
+    $self->_store_objects_in_chron($key_objs_hash, $date_obj, $optional_chron_args);
+    $self->_store_objects_in_cache($key_objs_hash) if $self->local_caching;
+    return 1;
+}
+
+sub _store_objects_in_cache {
+    my ($self, $key_objs_hash) = @_;
+    $self->{$_->key} = $_->value foreach (pairs %$key_objs_hash);
+    return 1;
+}
+
+sub _store_objects_in_chron {
+    my ($self, $key_objs_hash, $date_obj, $optional_chron_args) = @_;
+    my @atomic_write_pairs = pairmap { [$self->setting_namespace, $a, $b] } %$key_objs_hash;
+    $self->chronicle_writer->mset(\@atomic_write_pairs, $date_obj, @$optional_chron_args);
+    return 1;
+}
+
+=head2 get
+
+Takes either an arrayref of keys, gets them atomically, and returns a hashref of key->values,
+including the global_revision under the key '_global_revision'
+If a key has an empty value, it will return with undef.
+
+For convenience a get with just a key string will return the value only.
+
+Example:
+    get(['key1','key2','key3',...]);
+Returns:
+    {'key1' => 'value1', 'key2' => 'value2', 'key3' => 'value3',..., '_global_revision' => '<number>'}
+
+Example:
+    get(['key1']);
+Returns:
+    {'key1' => 'value1', '_global_revision' => '<number>'}
+
+Example:
+    get('key1');
+Returns:
+    'value1'
+
+=cut
+
+sub get {
+    my ($self, $keys) = @_;
+    my $single_get;
+
+    if (ref $keys eq '') {
```

Comment:
More traditionally written as

    if (not ref $keys)

or maybe

    unless(ref $keys)